_episodes_rmd/10-functions.Rmd

```diff
@@ -165,10 +167,110 @@ kelvin_to_celsius <- function(temp) {
 > {: .solution}
 {: .challenge}
 
+## Interlude: Defensive Programming
 
-We're going to define
-a function that calculates the Gross Domestic Product of a nation from the data
-available in our dataset:
+Now that we've begun to appreciate how writing functions provides an efficient
+way to make R code re-usable and modular, we should note that it is important
+to ensure that functions only work in their intended use-cases. This  is a part
+of the concept of _defensive programming_. Defensive programming encourages us
+to frequently check conditions and throw an error if something is
+wrong. These checks are referred to as assertion statements because we want to
+assert some condition is `TRUE` before proceeding. This makes it easier to
+debug because we have a better idea of where the error originated.
+
+### Checking conditions with `stopifnot`
+
+Let’s start by re-examining `fahr_to_kelvin`, our function for converting
+temperatures from Fahrenheit to Kelvin. It was defined like so:
+
+```{r}
+fahr_to_kelvin <- function(temp) {
+  kelvin <- ((temp - 32) * (5 / 9)) + 273.15
+  return(kelvin)
+}
+```
+
+For this function to work as intended, the argument `temp` must be a `numeric`
+value; otherwise, the mathematical procedure for converting between the two
+temperature scales will not work. To create an error, we can use the function
+`stop`. For example, since the arugment `temp` must be a `numeric` vector, we
+could check for this condition with an `if` statement and throw an error if the
+condition was violated. We could augment our function above like so:
+
+```{r}
+fahr_to_kelvin <- function(temp) {
+  if (!is.numeric(temp)) {
+    stop("temp must be a numeric vector.")
+  }
+  kelvin <- ((temp - 32) * (5 / 9)) + 273.15
+  return(kelvin)
+}
+```
+
+If we had multiple conditions or arguments to check, it would take many lines
+of code to check all of them. Luckily R provides the convenience function
+`stopifnot`. We can list as many requirements that should evaluate to `TRUE`;
```

Comment:
I'd add the `()` after each of these `stopifnot` references, so it is `stopifnot()`